### ISSUE TITLE
Sudden drop in conv3d_transpose GPU performance with large input sizes

### ISSUE THREAD

**Issue by mauriceqch** [tensorflow/tensorflow#25760]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): _Yes_
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): _Ubuntu 18.10_
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device: /
- TensorFlow installed from (source or binary): _Source_
- TensorFlow version (use command below): _'v1.12.0-0-ga6d8ffae09' 1.12.0_
- Python version: _3.6.7_
- Bazel version (if compiling from source): _0.18.1_
- GCC/Compiler version (if compiling from source): _4.8_
- CUDA/cuDNN version: _10.0, 7.3.1_
- CPU model: _Intel(R) Xeon(R) Bronze 3106 CPU @ 1.70GHz_
- GPU model and memory: _Nvidia GeForce GTX 1080 Ti 11GB_

**Describe the current behavior**
On GPU, the _conv3d_transpose_ operation shows a huge spike in terms of execution time when increasing input size from `(1, 128, 128, 128, 16)` to `(1, 256, 256, 256, 16)`.
This particular change makes average execution time without the first run change from **0.7s** to **92s**.
In comparison, the same operation on CPU goes from **2.3s** to **18.2s**.
This makes _conv3d_transpose_ perform better on CPU compared to CPU using this particular shape.

**Describe the expected behavior**
The _conv3d_transpose_ operation should not see such a sudden drop in performance when increasing input size.

**Code to reproduce the issue**
Attached is a python notebook that reproduces the issue.
[tf_conv3d.zip](https://github.com/tensorflow/tensorflow/files/2865718/tf_conv3d.zip)

**Other info / logs**
![download](https://user-images.githubusercontent.com/9284862/52798293-ee5fad80-3077-11e9-9b75-edc1ff504e8b.png)
This figure shows the average execution time of _conv3d_ and _conv3d_transpose_ operations on both CPU on GPU with a log scale for both x and y axis.
The x axis represents the resolution which determines the input shape in the following manner for _conv3d_ `(1, resolution, resolution, resolution, 16)` and determines the output shape for _conv3d_transpose_ `(1, resolution, resolution, resolution, 1)`.
GPU runs are in blue while CPU runs are in red.
_conv3d_ are represented with dash-dotted lines and _conv3d_tranpose_ with solid lines.
There is a sudden increase in the execution time of the _conv3d_transpose_ operation on GPU which goes from **0.7s** to **92s** when changing the resolution from **2^8** to **2^9**.

Attached is a full run of the attached python notebook that includes the full data.
[tf_conv3d.pdf](https://github.com/tensorflow/tensorflow/files/2865802/tf_conv3d.pdf)

**Comment by jvishnuvardhan** [comment]
@mauriceqch Could you try latest TF and check whether the bug persists or not? Thanks!

**Comment by mauriceqch** [comment]
@jvishnuvardhan Just tried on latest TF and the bug is still present.
In particular, I used the same setup as in the initial post with the following differences:

* TensorFlow version (use command below):  ~~_'v1.12.0-0-ga6d8ffae09' 1.12.0_~~ -> _b'v1.13.1-0-g6612da8951' 1.13.1_
* Bazel version (if compiling from source): ~~0.18.1~~ -> _0.21.0_

**Comment by sanjoy** [comment]
Hi @mauriceqch 

Thank you for reporting this.  Can you please minimize your reproducer and give us a simple Python file that prints out the various timings to the console?

**Comment by mauriceqch** [comment]
Hi @sanjoy 

Thank you for your interest !

Here is a minimal python file that prints timings to the console

```
import os
os.environ['TF_CPP_MIN_LOG_LEVEL'] = '3'
import tensorflow as tf
import numpy as np
import time

tf.get_logger().setLevel('ERROR')


kernel_size = (9, 9, 9)
strides = (2, 2, 2)
dtype = np.float32
batch_dim = 1
gpu_device = '/device:GPU:0'
cpu_device = '/cpu:0'
n_loops = 5
# Range of resolutions to test
min_res_exp = 5
max_res_exp = 10


def benchmark(sess, x, iters):
    times = np.zeros(iters)
    
    for i in range(iters):
        start_time = time.time()
        sess.run(x)
        wall_time = time.time() - start_time
        times[i] = wall_time
    
    ret = {
        'average': np.average(times),
        'std': np.std(times),
        'first': times[0],
        'average_without_first': np.average(times[1:]),
        'std_without_first': np.std(times[1:]),
        'iters': iters,
    }
    print(ret)
    return ret

def spatial_dim_iters(spatial_dim):
    return int((max_res_exp - np.log2(spatial_dim)) ** 2 * 3)

def run_conv3d(device, spatial_dim):
    tf.reset_default_graph()
    channels_dim = 4
    n_filters = 16
    sample_shape = (batch_dim, spatial_dim, spatial_dim, spatial_dim, channels_dim)
    sample = np.random.rand(*sample_shape).astype(dtype)
    print(f'Input shape: {sample.shape}')
    x_init = tf.placeholder(tf.float32, shape=sample_shape)
    x_val = tf.Variable(x_init)
    with tf.device(device):
        x = tf.layers.conv3d(x_val, n_filters, kernel_size, strides=strides, padding='same', data_format='channels_last')
    print(f'Layer: {x}')
    with tf.Session() as sess:
        sess.run(tf.global_variables_initializer(), feed_dict={x_init: sample})
        ret = benchmark(sess, x, spatial_dim_iters(spatial_dim))
        return ret


def run_conv3d_transpose(device, spatial_dim):
    tf.reset_default_graph()
    channels_dim = 16
    n_filters = 4
    # We divide spatial_dim by two as spatial_dim is the spatial dimension of the output
    sample_shape = (batch_dim, spatial_dim // 2, spatial_dim // 2, spatial_dim // 2, channels_dim)
    sample = np.random.rand(*sample_shape).astype(dtype)
    print(f'Input shape: {sample.shape}')
    x_init = tf.placeholder(tf.float32, shape=sample_shape)
    x_val = tf.Variable(x_init)
    with tf.device(device):
        x = tf.layers.conv3d_transpose(x_val, n_filters, kernel_size, strides=strides, padding='same', data_format='channels_last')
    print(f'Layer: {x}')
    with tf.Session() as sess:
        sess.run(tf.global_variables_initializer(), feed_dict={x_init: sample})
        ret = benchmark(sess, x, spatial_dim_iters(spatial_dim))
        return ret


CONV3D = 'CONV3D'
CONV3D_TRANSPOSE = 'CONV3D_TRANSPOSE'
methods = {CONV3D: run_conv3d, CONV3D_TRANSPOSE: run_conv3d_transpose}
test_setups = [
    [CONV3D, gpu_device],
    [CONV3D_TRANSPOSE, gpu_device],
    [CONV3D, cpu_device],
    [CONV3D_TRANSPOSE, cpu_device]
]
resolutions = [int(2**x) for x in range(min_res_exp, max_res_exp)]
print(f'Testing resolutions: {resolutions}\n')


results = []
for target, device in test_setups:
    for res in resolutions:
        print(f'Running test for: {target}, {device}, {res}')
        results.append([target, device, res, methods[target](device, res)])
        print('\n')

os.system('nvidia-smi')
print(tf.GIT_VERSION, tf.VERSION)
os.system('lscpu')
os.system('cat /usr/local/cuda/include/cudnn.h | grep CUDNN_MAJOR -A 2')
os.system('lsb_release -a')

```

**Comment by sanjoy** [comment]
@timshen91 PTAL.  Maybe we can extract out a test case for cuDNN that demonstrates this issue (assuming TF is using cuDNN correctly)?  /CC @nluehr

**Comment by kaixih** [comment]
I think this issue is mainly about slow conv3d_transpose with the shape of (1, 256, 256, 256, 16).

On my GV100 GPU, I can get 115ms for the (1, 128, 128, 128, 16) but 44762ms for the (1, 256, 256, 256, 16).

After some digging, I found in the first case (1, 128, 128, 128, 16) the op chose the algo0 (115ms) over the algo1 (5643ms) during the autotuning phase. Whereas, in the second case (1, 256, 256, 256, 16), the op could only pick algo1 (44762ms) since the algo0 is not supported (insufficient workspace).

@mauriceqch Can I ask what do you need this case for? For benchmarking or real models?

**Comment by mauriceqch** [comment]
@kaixih Many thanks for your in-depth investigation. 
I am using this for point cloud compression applications, specifically, the geometry information.
Currently, I am working around this issue by splitting point clouds into blocks.

**Comment by tensorflowbutler** [comment]
Hi There,

 We are checking to see if you still need help on this issue, as you are using an older version of tensorflow(1.x) which is officially considered as end of life. We recommend that you upgrade to 2.4 or later version and let us know if the issue still persists in newer versions. 

 This issue will be closed automatically 7 days from now. If you still need help with this issue, Please open a new issue for any help you need against 2.x, and we will get you the right help.
